File: Sistemas-Operacionais-de Tempo real/Sistemas-Operacionais-Embarcados-main/supervisor.c
/*
 * supervisor.c
 * Pai cria filhos e monitora seus estados em tempo real
 *
 * Compilação:
 *   gcc supervisor.c -o supervisor
 *
 * Execução:
 *   ./supervisor
 *
 * Em outro terminal, você pode enviar sinais:
 *   kill -STOP <PID_DO_FILHO>   // coloca filho em T (Stopped)
 *   kill -CONT <PID_DO_FILHO>   // retoma filho
 *   kill -TERM <PID_DO_FILHO>   // encerra filho (vai gerar evento)
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <sys/wait.h>

int main() {
    pid_t pid1, pid2;

    // Cria primeiro filho
    pid1 = fork();
    if (pid1 == 0) {
        // Código do filho 1: loop infinito
        while (1) {
            pause(); // aguarda sinais
        }
        exit(0);
    }

    // Cria segundo filho
    pid2 = fork();
    if (pid2 == 0) {
        // Código do filho 2: loop infinito
        while (1) {
            pause(); // aguarda sinais
        }
        exit(0);
    }

    printf("Pai supervisor PID=%d criou filhos: %d e %d\n", getpid(), pid1, pid2);

    // Loop de supervisão
    while (1) {
        int status;
        pid_t w = waitpid(-1, &status, WUNTRACED | WCONTINUED);

        if (w == -1) {
            perror("waitpid");
            exit(1);
        }

        if (WIFEXITED(status)) {
            printf("Filho %d terminou normalmente com código %d\n", w, WEXITSTATUS(status));
        } 
        else if (WIFSIGNALED(status)) {
            printf("Filho %d morreu por sinal %d\n", w, WTERMSIG(status));
        } 
        else if (WIFSTOPPED(status)) {
            printf("Filho %d foi PARADO (SIGSTOP ou Ctrl+Z), sinal=%d\n", w, WSTOPSIG(status));
        } 
        else if (WIFCONTINUED(status)) {
            printf("Filho %d foi CONTINUADO (SIGCONT)\n", w);
        }
    }

    return 0;
}

File: Sistemas-Operacionais-de Tempo real/Sistemas-Operacionais-Embarcados-main/supervisor.py
"""
supervisor.py
Pai cria filhos e monitora seus estados em tempo real

Execução:
    python3 supervisor.py

Em outro terminal, você pode enviar sinais:
    kill -STOP <PID_FILHO>   # coloca o filho em estado T (Stopped)
    kill -CONT <PID_FILHO>   # retoma filho
    kill -TERM <PID_FILHO>   # encerra filho
"""

import os
import signal

def filho():
    # Cada filho fica em loop infinito, apenas aguardando sinais
    print(f"Filho iniciado, PID={os.getpid()}")
    while True:
        signal.pause()  # bloqueia até receber um sinal

def pai_supervisor():
    # Cria primeiro filho
    pid1 = os.fork()
    if pid1 == 0:
        filho()
        os._exit(0)

    # Cria segundo filho
    pid2 = os.fork()
    if pid2 == 0:
        filho()
        os._exit(0)

    print(f"Pai supervisor PID={os.getpid()} criou filhos: {pid1}, {pid2}")

    # Loop de supervisão
    while True:
        # WUNTRACED = 2, WCONTINUED = 8 (valores das flags no Linux)
        wpid, status = os.waitpid(-1, os.WUNTRACED | os.WCONTINUED)

        if os.WIFEXITED(status):
            print(f"Filho {wpid} terminou normalmente com código {os.WEXITSTATUS(status)}")
        elif os.WIFSIGNALED(status):
            print(f"Filho {wpid} morreu por sinal {os.WTERMSIG(status)}")
        elif os.WIFSTOPPED(status):
            print(f"Filho {wpid} foi PARADO (SIGSTOP ou Ctrl+Z), sinal={os.WSTOPSIG(status)}")
        elif os.WIFCONTINUED(status):
            print(f"Filho {wpid} foi CONTINUADO (SIGCONT)")

if __name__ == "__main__":
    pai_supervisor()
